src/main/scala/io/iohk/ethereum/blockchain/sync/regular/BlockImporter.scala

```diff
@@ -0,0 +1,314 @@
+
+    importWith(
+      state, {
+        doLog(importMessages.preImport())
+        ledger
+          .importBlock(block)(context.dispatcher)
+          .tap {
+            case BlockImportedToTop(importedBlocksData) =>
+              val (blocks, receipts) = importedBlocksData.map(data => (data.block, data.td)).unzip
+              broadcastBlocks(blocks, receipts)
+              updateTxAndOmmerPools(importedBlocksData.map(_.block), Seq.empty)
+              doLog(importMessages.importedToTheTop())
+
+            case BlockEnqueued =>
+              doLog(importMessages.enqueued())
+              ommersPool ! AddOmmers(block.header)
+
+            case DuplicateBlock =>
+              doLog(importMessages.duplicated())
+
+            case UnknownParent =>
+              // This is normal when receiving broadcast blocks
+              doLog(importMessages.orphaned())
+
+            case ChainReorganised(oldBranch, newBranch, totalDifficulties) =>
+              updateTxAndOmmerPools(newBranch, oldBranch)
+              broadcastBlocks(newBranch, totalDifficulties)
+              doLog(importMessages.reorganisedChain(newBranch))
+
+            case BlockImportFailed(error) =>
+              doLog(importMessages.importFailed(error))
+              if (informFetcherOnFail) {
+                fetcher ! BlockFetcher.BlockImportFailed(Block.number(block), error)
+              }
+          }
+          .map(_ => running _)
+          .recover {
+            case missingNodeEx: MissingNodeException if syncConfig.redownloadMissingStateNodes =>
+              // state node re-download will be handled when downloading headers
+              doLog(importMessages.missingStateNode(missingNodeEx))
+              running
+          }
+      }
+    )
+  }
+
+  private def broadcastBlocks(blocks: List[Block], totalDifficulties: List[BigInt]): Unit = {
+    val newBlocks = (blocks, totalDifficulties).mapN(NewBlock.apply)
+    broadcastNewBlocks(newBlocks)
+  }
+
+  private def broadcastNewBlocks(blocks: List[NewBlock]): Unit = {
+    broadcaster ! BroadcastBlocks(blocks)
+  }
+
+  private def updateTxAndOmmerPools(blocksAdded: Seq[Block], blocksRemoved: Seq[Block]): Unit = {
+    blocksRemoved.headOption.foreach(block => ommersPool ! AddOmmers(block.header))
+    blocksRemoved.foreach(block => pendingTransactionsManager ! AddTransactions(block.body.transactionList.toSet))
+
+    blocksAdded.foreach { block =>
+      ommersPool ! RemoveOmmers(block.header :: block.body.uncleNodesList.toList)
+      pendingTransactionsManager ! RemoveTransactions(block.body.transactionList)
+    }
+  }
+
+  private def importWith(state: ImporterState, importFuture: => Future[Behavior]): Unit = {
+    val newState = ImporterState.importingBlocks(state)
+    context become running(newState)
+    importFuture.onComplete {
+      case Failure(ex) => throw ex
+      case Success(behavior) => context become behavior(ImporterState.notImportingBlocks(newState))
```

Comment:
You shouldn't switch context in `onComplete`. It isn't processed by the same thread.

From Akka documentation:
```
/**
...
   * WARNING: Only valid within the Actor itself, so do not close over it and
   * publish it to other threads!
...
  implicit val context: ActorContext
```

Reply:
Good catch